src/TraverseTracks.cpp

```diff
@@ -0,0 +1,153 @@
+#include "TraverseTracks.h"
+
+/**
+ * @brief Constructor for the TravseSegments class assigns the TrackGenerator
+ *        and pulls relevant information from it.
+ */
+TraverseTracks::TraverseTracks(TrackGenerator* track_generator) {
+
+  /* Save the track generator */
+  _track_generator = track_generator;
+
+  /* Determine the type of segment formation used */
+  _segment_formation = track_generator->getSegmentFormation();
+}
+
+
+/**
+ * @brief Destructor for TraverseTracks
+ */
+TraverseTracks::~TraverseTracks() {
+}
+
+
+/**
+ * @breif Loops over Tracks, applying the provided kernels to all segments and
+ *        the functionality described in onTrack(...) to all Tracks.
+ * @details The segment formation method imported from the TrackGenerator
+ *          during construction is used to redirect to the appropriate looping
+ *          scheme. If kernels are provided (not NULL) then they are deleted at
+ *          the end of the looping scheme.
+ * @param kernels MOCKernels to apply to all segments
+ */
+void TraverseTracks::loopOverTracks(MOCKernel** kernels) {
+
+  switch (_segment_formation) {
+    case EXPLICIT_2D:
+      loopOverTracks2D(kernels);
+      break;
+  }
+
+  if (kernels != NULL) {
+    delete kernels[0];
+    delete [] kernels;
+  }
+}
+
+
+/**
+ * @breif Loops over Tracks by parallel groups to avoid potential conflicts,
+ *        applying the provided kernels to all segments and the functionality
+ *        described in onTrack(...) to all Tracks.
+ * @details The segment formation method imported from the TrackGenerator
+ *          during construction is used to redirect to the appropriate looping
+ *          scheme. If kernels are provided (not NULL) then they are deleted at
+ *          the end of the looping scheme.
+ * @param kernels MOCKernels to apply to all segments
+ */
+void TraverseTracks::loopOverTracksByParallelGroup(MOCKernel** kernels) {
+  loopOverTracksByParallelGroup2D(kernels);
+}
+
+
+/**
+ * @brief Loops over all explicit 2D Tracks
+ * @details The onTrack(...) function is applied to all 2D Tracks and the
```

Comment:
You might add to you `@details` comment block that this `loopOverTracks2D(...)` only differs from `loopOverTracksByParallelGroup2D(...)` in terms of the parallelization (unless I've missed something). If I understand things correctly, `loopOverTracks2D(...)` does not enforce any chunking or ordering of operations across tracks, while `loopOverTracksByParallelGroup2D(...)` does enforce chunking / ordering of operations according to some angular criteria. This should be added to the comments for each method to really pinpoint what differentiates them.